source/common/router/config_impl.cc

```diff
@@ -163,18 +201,44 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost,
       priority_(ConfigUtility::parsePriority(route.route().priority())),
       request_headers_parser_(RequestHeaderParser::parse(route.route().request_headers_to_add())),
       opaque_config_(parseOpaqueConfig(route)), decorator_(parseDecorator(route)) {
+  if (route.route().has_metadata_match()) {
+    const auto filter_it = route.route().metadata_match().filter_metadata().find(
+        Envoy::Config::MetadataFilters::get().ENVOY_LB);
+    if (filter_it != route.route().metadata_match().filter_metadata().end()) {
+      metadata_match_criteria_.reset(new MetadataMatchCriteriaImpl(filter_it->second));
+    }
+  }
+
   // If this is a weighted_cluster, we create N internal route entries
   // (called WeightedClusterEntry), such that each object is a simple
-  // single cluster, pointing back to the parent.
+  // single cluster, pointing back to the parent. Metadata criteria
+  // from the weighted cluster (if any) are merged with and override
+  // the criteria from the route.
```

Comment:
Quick question: isn't route more specific than cluster, so would it not make sense to have the merge/override the other way around? I don't have any clue here, just asking.

Reply:
I think the argument can made either way. 

The way I arrived at this is a request is mapped to a Route, which further maps the request to a Cluster and that it makes sense for the metadata criteria to be built up in the same order.

That said, we don't plan on depending on this behavior (we may use both fields, but the sets of keys in each will be disjoint).

Reply:
OK, I have no strong opinion here, was just curious.